File: neoforge/src/main/java/generations/gg/generations/core/generationscore/forge/datagen/generators/blocks/UltraBlockModelDataGen.java
package generations.gg.generations.core.generationscore.forge.datagen.generators.blocks;

import generations.gg.generations.core.generationscore.common.GenerationsCore;
import generations.gg.generations.core.generationscore.common.world.level.block.set.GenerationsUltraBlockSet;
import net.minecraft.data.BlockFamily;
import net.minecraft.world.level.block.*;
import net.neoforged.neoforge.client.model.generators.BlockModelBuilder;

public class UltraBlockModelDataGen extends GenerationsBlockStateProvider.Proxied {

	public UltraBlockModelDataGen(GenerationsBlockStateProvider provider) {
		super(provider);
	}

	@Override
	public void registerStatesAndModels() {
		GenerationsUltraBlockSet.ultraBlockSets.forEach(arg -> registerBlockFamily(arg.blockFamily));
	}
	private void registerBlockFamily(BlockFamily family) {
		registerUltraBlock(family.getBaseBlock());
		family.getVariants().keySet().forEach(variant -> proccessVariant(variant, family));
	}

	private void proccessVariant(BlockFamily.Variant variant, BlockFamily family) {
		Block original = family.getBaseBlock();
		Block variantTarget = family.getVariants().get(variant);
		switch (variant) {
			case BUTTON -> registerUltraButton((ButtonBlock) variantTarget, original);
			case SLAB -> registerUltraSlab((SlabBlock) variantTarget, original);
			case STAIRS -> registerUltraStairs((StairBlock) variantTarget, original);
			case WALL -> registerUltraWall((WallBlock) variantTarget, original);
			case PRESSURE_PLATE -> registerUltraPressurePlate((PressurePlateBlock) variantTarget, original);
		}
	}

	private void registerUltraBlock(Block block) {
		String name = block.getDescriptionId().replace("block.generations_core.", "");
		simpleBlockWithItem(block, models().singleTexture(name, GenerationsCore.id("block/ultra/ultra_block"), "all", GenerationsCore.id("block/ultra/" + name)));
	}
	private void registerUltraStairs(StairBlock stair, Block block) {
		String name = block.getDescriptionId().replace("block.generations_core.", "");
		String stairName = stair.getDescriptionId().replace("block.generations_core.", "");
		BlockModelBuilder ultrastair = models().singleTexture(stairName, GenerationsCore.id("block/ultra/ultra_stairs"), "bottom", GenerationsCore.id("block/ultra/" + name)).texture("top", GenerationsCore.id("block/ultra/" + name)).texture("side", GenerationsCore.id("block/ultra/" + name));
		BlockModelBuilder ultrastairInner = models().singleTexture(stairName + "_inner", GenerationsCore.id("block/ultra/ultra_stairs_inner"), "bottom", GenerationsCore.id("block/ultra/" + name)).texture("side", GenerationsCore.id("block/ultra/" + name)).texture("top", GenerationsCore.id("block/ultra/" + name));
		BlockModelBuilder ultrastairOuter = models().singleTexture(stairName + "_outer", GenerationsCore.id("block/ultra/ultra_stairs_outer"), "bottom", GenerationsCore.id("block/ultra/" + name)).texture("side", GenerationsCore.id("block/ultra/" + name)).texture("top", GenerationsCore.id("block/ultra/" + name));
		stairsBlock(stair, ultrastair, ultrastairInner, ultrastairOuter);
		simpleBlockItem(stair, ultrastair);
	}

	private void registerUltraSlab(SlabBlock slab, Block block) {
		String name = block.getDescriptionId().replace("block.generations_core.", "");
		String slabname = slab.getDescriptionId().replace("block.generations_core.", "");
		BlockModelBuilder ultraSlab = models().singleTexture(slabname, GenerationsCore.id("block/ultra/ultra_slab"), "bottom", GenerationsCore.id("block/ultra/" + name)).texture("top", GenerationsCore.id("block/ultra/" + name)).texture("side", GenerationsCore.id("block/ultra/" + name));
		BlockModelBuilder ultraSlabTop = models().singleTexture(slabname + "_top", GenerationsCore.id("block/ultra/ultra_slab_top"), "bottom", GenerationsCore.id("block/ultra/" + name)).texture("top", GenerationsCore.id("block/ultra/" + name)).texture("side", GenerationsCore.id("block/ultra/" + name));
		BlockModelBuilder ultraSlabDouble = models().singleTexture(slabname + "_double", GenerationsCore.id("block/ultra/ultra_slab_double"), "end", GenerationsCore.id("block/ultra/" + name)).texture("side", GenerationsCore.id("block/ultra/" + name));
		slabBlock(slab, ultraSlab, ultraSlabTop, ultraSlabDouble);
		simpleBlockItem(slab, ultraSlab);
	}

	private void registerUltraButton(ButtonBlock button, Block block) {
		String name = block.getDescriptionId().replace("block.generations_core.", "");
		String buttonname = button.getDescriptionId().replace("block.generations_core.", "");
		BlockModelBuilder ultraBlock = models().singleTexture(buttonname, GenerationsCore.id("block/ultra/ultra_button"), "texture", GenerationsCore.id("block/ultra/" + name));
		BlockModelBuilder ultraButtonPressed = models().singleTexture(buttonname + "_pressed", GenerationsCore.id("block/ultra/ultra_button_pressed"), "texture", GenerationsCore.id("block/ultra/" + name));
		buttonBlock(button, ultraBlock, ultraButtonPressed);
		itemModels().buttonInventory(buttonname, GenerationsCore.id("block/ultra/" + name));
	}

	private void registerUltraWall(WallBlock wall, Block block) {
		String name = block.getDescriptionId().replace("block.generations_core.", "");
		String wallname = wall.getDescriptionId().replace("block.generations_core.", "");
		BlockModelBuilder ultraWallPost = models().singleTexture(wallname + "_post", GenerationsCore.id("block/ultra/ultra_wall_post"), "wall", GenerationsCore.id("block/ultra/" + name));
		BlockModelBuilder ultraWallSide = models().singleTexture(wallname + "_side", GenerationsCore.id("block/ultra/ultra_wall_side"), "wall", GenerationsCore.id("block/ultra/" + name));
		BlockModelBuilder ultraWallSideTall = models().singleTexture(wallname + "_side_tall", GenerationsCore.id("block/ultra/ultra_wall_side_tall"), "wall", GenerationsCore.id("block/ultra/" + name));
		wallBlock(wall, ultraWallPost, ultraWallSide, ultraWallSideTall);
		itemModels().wallInventory(wallname, GenerationsCore.id("block/ultra/" + name));
	}

	private void registerUltraPressurePlate(PressurePlateBlock pressurePlate, Block block) {
		String name = block.getDescriptionId().replace("block.generations_core.", "");
		String pressureplatename = pressurePlate.getDescriptionId().replace("block.generations_core.", "");
		BlockModelBuilder ultraPressurePlateUp = models().singleTexture(pressureplatename + "_up", GenerationsCore.id("block/ultra/ultra_pressure_plate_up"), "texture", GenerationsCore.id("block/ultra/" + name));
		BlockModelBuilder ultraPressurePlateDown = models().singleTexture(pressureplatename + "_down", GenerationsCore.id("block/ultra/ultra_pressure_plate_down"), "texture", GenerationsCore.id("block/ultra/" + name));
		pressurePlateBlock(pressurePlate, ultraPressurePlateUp, ultraPressurePlateDown);
		itemModels().pressurePlate(pressureplatename, GenerationsCore.id("block/ultra/" + name));
	}
}

File: src/neoforge/main/kotlin/generations/gg/generations/core/generationscore/forge/datagen/generators/blocks/UltraBlockModelDataGen.kt
package generations.gg.generations.core.generationscore.forge.datagen.generators.blocks

import generations.gg.generations.core.generationscore.common.GenerationsCore.id
import generations.gg.generations.core.generationscore.common.world.level.block.set.GenerationsUltraBlockSet
import net.minecraft.data.BlockFamily
import net.minecraft.world.level.block.*
import java.util.function.Consumer

class UltraBlockModelDataGen(provider: GenerationsBlockStateProvider) :
    GenerationsBlockStateProvider.Proxied(provider) {
    override fun registerStatesAndModels() {
        GenerationsUltraBlockSet.ultraBlockSets.forEach(Consumer { arg: GenerationsUltraBlockSet ->
            registerBlockFamily(
                arg.blockFamily!!
            )
        })
    }

    private fun registerBlockFamily(family: BlockFamily) {
        registerUltraBlock(family.baseBlock)
        family.variants.keys.forEach(Consumer { variant: BlockFamily.Variant -> proccessVariant(variant, family) })
    }

    private fun proccessVariant(variant: BlockFamily.Variant, family: BlockFamily) {
        val original = family.baseBlock
        val variantTarget = family.variants[variant]
        when (variant) {
            BlockFamily.Variant.BUTTON -> registerUltraButton((variantTarget as ButtonBlock), original)
            BlockFamily.Variant.SLAB -> registerUltraSlab((variantTarget as SlabBlock), original)
            BlockFamily.Variant.STAIRS -> registerUltraStairs((variantTarget as StairBlock), original)
            BlockFamily.Variant.WALL -> registerUltraWall((variantTarget as WallBlock), original)
            BlockFamily.Variant.PRESSURE_PLATE -> registerUltraPressurePlate((variantTarget as PressurePlateBlock), original)
            else -> {}
        }
    }

    private fun registerUltraBlock(block: Block) {
        val name = block.descriptionId.replace("block.generations_core.", "")
        simpleBlockWithItem(
            block, models().singleTexture(
                name, id("block/ultra/ultra_block"), "all", id(
                    "block/ultra/$name"
                )
            )
        )
    }

    private fun registerUltraStairs(stair: StairBlock, block: Block) {
        val name = block.descriptionId.replace("block.generations_core.", "")
        val stairName = stair.descriptionId.replace("block.generations_core.", "")
        val ultrastair = models().singleTexture(
            stairName, id("block/ultra/ultra_stairs"), "bottom", id(
                "block/ultra/$name"
            )
        ).texture("top", id("block/ultra/$name")).texture("side", id("block/ultra/$name"))
        val ultrastairInner = models().singleTexture(
            stairName + "_inner", id("block/ultra/ultra_stairs_inner"), "bottom", id(
                "block/ultra/$name"
            )
        ).texture("side", id("block/ultra/$name")).texture("top", id("block/ultra/$name"))
        val ultrastairOuter = models().singleTexture(
            stairName + "_outer", id("block/ultra/ultra_stairs_outer"), "bottom", id(
                "block/ultra/$name"
            )
        ).texture("side", id("block/ultra/$name")).texture("top", id("block/ultra/$name"))
        stairsBlock(stair, ultrastair, ultrastairInner, ultrastairOuter)
        simpleBlockItem(stair, ultrastair)
    }

    private fun registerUltraSlab(slab: SlabBlock, block: Block) {
        val name = block.descriptionId.replace("block.generations_core.", "")
        val slabname = slab.descriptionId.replace("block.generations_core.", "")
        val ultraSlab = models().singleTexture(
            slabname, id("block/ultra/ultra_slab"), "bottom", id(
                "block/ultra/$name"
            )
        ).texture("top", id("block/ultra/$name")).texture("side", id("block/ultra/$name"))
        val ultraSlabTop = models().singleTexture(
            slabname + "_top", id("block/ultra/ultra_slab_top"), "bottom", id(
                "block/ultra/$name"
            )
        ).texture("top", id("block/ultra/$name")).texture("side", id("block/ultra/$name"))
        val ultraSlabDouble = models().singleTexture(
            slabname + "_double", id("block/ultra/ultra_slab_double"), "end", id(
                "block/ultra/$name"
            )
        ).texture("side", id("block/ultra/$name"))
        slabBlock(slab, ultraSlab, ultraSlabTop, ultraSlabDouble)
        simpleBlockItem(slab, ultraSlab)
    }

    private fun registerUltraButton(button: ButtonBlock, block: Block) {
        val name = block.descriptionId.replace("block.generations_core.", "")
        val buttonname = button.descriptionId.replace("block.generations_core.", "")
        val ultraBlock = models().singleTexture(
            buttonname, id("block/ultra/ultra_button"), "texture", id(
                "block/ultra/$name"
            )
        )
        val ultraButtonPressed = models().singleTexture(
            buttonname + "_pressed", id("block/ultra/ultra_button_pressed"), "texture", id(
                "block/ultra/$name"
            )
        )
        buttonBlock(button, ultraBlock, ultraButtonPressed)
        itemModels().buttonInventory(buttonname, id("block/ultra/$name"))
    }

    private fun registerUltraWall(wall: WallBlock, block: Block) {
        val name = block.descriptionId.replace("block.generations_core.", "")
        val wallname = wall.descriptionId.replace("block.generations_core.", "")
        val ultraWallPost = models().singleTexture(
            wallname + "_post", id("block/ultra/ultra_wall_post"), "wall", id(
                "block/ultra/$name"
            )
        )
        val ultraWallSide = models().singleTexture(
            wallname + "_side", id("block/ultra/ultra_wall_side"), "wall", id(
                "block/ultra/$name"
            )
        )
        val ultraWallSideTall = models().singleTexture(
            wallname + "_side_tall", id("block/ultra/ultra_wall_side_tall"), "wall", id(
                "block/ultra/$name"
            )
        )
        wallBlock(wall, ultraWallPost, ultraWallSide, ultraWallSideTall)
        itemModels().wallInventory(wallname, id("block/ultra/$name"))
    }

    private fun registerUltraPressurePlate(pressurePlate: PressurePlateBlock, block: Block) {
        val name = block.descriptionId.replace("block.generations_core.", "")
        val pressureplatename = pressurePlate.descriptionId.replace("block.generations_core.", "")
        val ultraPressurePlateUp = models().singleTexture(
            pressureplatename + "_up", id("block/ultra/ultra_pressure_plate_up"), "texture", id(
                "block/ultra/$name"
            )
        )
        val ultraPressurePlateDown = models().singleTexture(
            pressureplatename + "_down", id("block/ultra/ultra_pressure_plate_down"), "texture", id(
                "block/ultra/$name"
            )
        )
        pressurePlateBlock(pressurePlate, ultraPressurePlateUp, ultraPressurePlateDown)
        itemModels().pressurePlate(pressureplatename, id("block/ultra/$name"))
    }
}
